File: tests/test-data/workspace/display_options_test.py
"""
Test file for display options feature testing.
Creates various types of images that are designed to make display option effects clearly visible.
"""
import numpy as np


def create_rgb_gradient():
    """
    Create an RGB image with distinct color regions.
    Left third is red, middle third is green, right third is blue.
    This makes it easy to verify R/G/B channel filtering.
    """
    img = np.zeros((100, 150, 3), dtype=np.uint8)
    # Red region (left)
    img[:, :50, 0] = 255
    # Green region (middle)
    img[:, 50:100, 1] = 255
    # Blue region (right)
    img[:, 100:, 2] = 255
    return img


def create_grayscale_gradient():
    """
    Create a grayscale gradient from black to white.
    Good for testing high contrast and invert options.
    """
    gradient = np.linspace(0, 255, 200, dtype=np.uint8)
    img = np.tile(gradient, (100, 1))
    return img


def create_segmentation_image():
    """
    Create a segmentation mask with distinct integer labels.
    Uses values 0-5 to represent different segments.
    """
    img = np.zeros((100, 150), dtype=np.uint8)
    # Create 6 horizontal bands with different labels
    for i in range(6):
        img[i * 16:(i + 1) * 16, :] = i
    # Add some variety with vertical sections
    img[:, 75:100] = img[:, 75:100] + 3
    img = img % 6  # Keep values in 0-5 range
    return img


def create_heatmap_image():
    """
    Create a float32 image suitable for heatmap visualization.
    Uses a 2D Gaussian-like pattern with values from 0 to 1.
    """
    x = np.linspace(-2, 2, 150)
    y = np.linspace(-2, 2, 100)
    X, Y = np.meshgrid(x, y)
    img = np.exp(-(X**2 + Y**2) / 2).astype(np.float32)
    return img


def create_rgba_image():
    """
    Create an RGBA image with a checkerboard alpha pattern.
    Good for testing ignore alpha option.
    """
    img = np.zeros((100, 100, 4), dtype=np.uint8)
    # Fill with a gradient color
    img[:, :, 0] = 200  # Red
    img[:, :, 1] = 100  # Green
    img[:, :, 2] = 50   # Blue
    # Create checkerboard alpha pattern
    for i in range(10):
        for j in range(10):
            if (i + j) % 2 == 0:
                img[i*10:(i+1)*10, j*10:(j+1)*10, 3] = 255
            else:
                img[i*10:(i+1)*10, j*10:(j+1)*10, 3] = 128
    return img


def create_bgr_test_image():
    """
    Create an image where BGR vs RGB swap is clearly visible.
    Creates cyan on one side (should become red-ish when swapped)
    and red on the other (should become cyan when swapped).
    """
    img = np.zeros((100, 100, 3), dtype=np.uint8)
    # Left half: Cyan (G+B) - will become Yellow (R+G) when BGR swapped
    img[:, :50, 1] = 255
    img[:, :50, 2] = 255
    # Right half: Red - will become Blue when BGR swapped
    img[: , 50:, 0] = 255
    return img


def create_high_contrast_test():
    """
    Create a low contrast grayscale image.
    Values range from 100-155 to test high contrast enhancement.
    """
    gradient = np.linspace(100, 155, 200, dtype=np.uint8)
    img = np.tile(gradient, (100, 1))
    return img


def main():
    # Create all test images
    rgb_gradient = create_rgb_gradient()
    grayscale = create_grayscale_gradient()
    segmentation = create_segmentation_image()
    heatmap = create_heatmap_image()
    rgba = create_rgba_image()
    bgr_test = create_bgr_test_image()
    low_contrast = create_high_contrast_test()

    # Breakpoint for testing - all images are available here
    breakpoint()

    print("Display options test images created successfully")


if __name__ == "__main__":
    main()

File: tests/ui-test/display-options.test.ts
/**
 * Display Options Test Suite
 *
 * Tests the display options feature by creating various types of images
 * and verifying that the display options buttons work correctly.
 *
 * Test images created:
 * - RGB gradient: Red/Green/Blue regions for channel filtering tests
 * - Grayscale gradient: For high contrast and invert tests
 * - Segmentation mask: Integer labels for segmentation colormap test
 * - Heatmap image: Float32 gaussian for heatmap colormap test
 * - RGBA image: Checkerboard alpha for ignore alpha test
 * - BGR test image: Cyan/Red pattern for RGB/BGR swap test
 * - Low contrast image: For high contrast enhancement test
 */

import type { WebDriver } from 'vscode-extension-tester';
import { VSBrowser } from 'vscode-extension-tester';
import { DebugTestHelper } from './DebugTestHelper';
import { fileInWorkspace, openWorkspace } from './globals';
import {
  assertBrighterThan,
  assertChannelSwapped,
  assertDominantChannel,
  clickDisplayOption,
  sampleRegion,
} from './image-verification-utils';

describe('display options tests', () => {
  let debugHelper: DebugTestHelper;
  let driver: WebDriver;

  before(async () => {
    DebugTestHelper.logger.step('Opening workspace for display options tests');
    await openWorkspace();
    DebugTestHelper.logger.step('Workspace opened');

    driver = VSBrowser.instance.driver;

    // Initialize the debug helper
    debugHelper = DebugTestHelper.getInstance({
      timeout: 30000,
      retryCount: 5,
      sleepDuration: 1000,
    });
  }).timeout(60000);

  afterEach(async () => {
    // Clean up after tests
    if (debugHelper) {
      await debugHelper.cleanup();
    }
  }).timeout(60000);

  after(async () => {
    DebugTestHelper.reset();
  }).timeout(5000);

  async function viewVariable(variableName: string): Promise<void> {
    await debugHelper.performVariableAction({
      variableName,
      actionLabel: 'View Image',
      retrySetup: true,
      setupRetries: 5,
      type: 'variable',
    });
    await debugHelper.wait(1000);
    await debugHelper.getWebviewEditor();
    await debugHelper.wait(500);
  }

  /**
   * Comprehensive test that tests all display options in a single debug session.
   * This is more reliable than running separate tests.
   */
  it('should test all display options', async () => {
    debugHelper.setCurrentTest('display-options');

    DebugTestHelper.logger.testStart('Testing all display options');

    // Setup debugging with the display options test file
    await debugHelper.setupEditorForDebug({
      fileName: fileInWorkspace('display_options_test.py'),
      debugConfig: 'Python: Current File',
      openFile: true,
    });

    await debugHelper.startDebugging();
    await debugHelper.waitForBreakpoint();
    await debugHelper.wait(2000); // extra wait for extension to process debug state in CI

    await debugHelper.expandImageWatchSection();
    await debugHelper.collapseOtherDebugSections();

    // Explicitly trigger run-setup to ensure the extension detects variables
    // (important on first test run in CI where extension may still be initializing)
    await debugHelper.runSetup();
    await debugHelper.wait(1000);

    // ===== Test 1: RGB Channel Filters =====
    DebugTestHelper.logger.step('Testing RGB channel display options on rgb_gradient...');

    await viewVariable('rgb_gradient');

    // Test Red channel filter
    const redClicked = await clickDisplayOption(driver, 'Red Channel');
    if (redClicked) {
      DebugTestHelper.logger.success('Red channel filter applied');

      // Verify pixel rendering: left region (pure red) should remain red;
      // middle region (pure green) should become dark since its red content is zero.
      const imgAfterRed = await debugHelper.captureCanvasImage();
      if (imgAfterRed) {
        const leftAfterRed = sampleRegion(imgAfterRed, 0.05, 0.2, 0.25, 0.6);
        const middleAfterRed = sampleRegion(imgAfterRed, 0.38, 0.2, 0.25, 0.6);
        DebugTestHelper.logger.info(`Red filter — left: ${JSON.stringify(leftAfterRed)}, middle: ${JSON.stringify(middleAfterRed)}`);
        assertDominantChannel(leftAfterRed, 'r', 40, 'left region after red channel filter');
        assertBrighterThan(leftAfterRed, middleAfterRed, 40, 'left/red brighter than middle/green after red filter');
      }
    }

    // Reset before next test
    await clickDisplayOption(driver, 'Reset');

    // Test Green channel filter
    const greenClicked = await clickDisplayOption(driver, 'Green Channel');
    if (greenClicked) {
      DebugTestHelper.logger.success('Green channel filter applied');
    }

    // Reset before next test
    await clickDisplayOption(driver, 'Reset');

    // Test Blue channel filter
    const blueClicked = await clickDisplayOption(driver, 'Blue Channel');
    if (blueClicked) {
      DebugTestHelper.logger.success('Blue channel filter applied');
    }

    // Reset and test Grayscale
    await clickDisplayOption(driver, 'Reset');
    await clickDisplayOption(driver, 'Grayscale');

    // ===== Test 2: Grayscale and Contrast Options =====
    DebugTestHelper.logger.step('Testing grayscale and contrast display options...');

    // View the grayscale gradient
    await viewVariable('grayscale');

    // Test Invert
    await clickDisplayOption(driver, 'Invert Colors');

    // Reset
    await clickDisplayOption(driver, 'Reset');

    // View the low contrast image
    await viewVariable('low_contrast');

    // Test High Contrast
    await clickDisplayOption(driver, 'High Contrast');

    // ===== Test 3: Heatmap and Segmentation Colormaps =====
    DebugTestHelper.logger.step('Testing heatmap and segmentation display options...');

    // View the heatmap image (float32 gaussian)
    await viewVariable('heatmap');

    // Test Heatmap colormap
    await clickDisplayOption(driver, 'Heatmap');

    // View the segmentation image
    await viewVariable('segmentation');

    // Test Segmentation colormap
    await clickDisplayOption(driver, 'Segmentation');

    // ===== Test 4: RGBA and BGR Options =====
    DebugTestHelper.logger.step('Testing RGBA and BGR display options...');

    // View the RGBA image
    await viewVariable('rgba');

    // Test Ignore Alpha
    await clickDisplayOption(driver, 'Ignore Alpha');

    // View the BGR test image — capture BEFORE swap for comparison
    await viewVariable('bgr_test');
    const imgBeforeBgr = await debugHelper.captureCanvasImage();

    // Test Swap RGB/BGR
    await clickDisplayOption(driver, 'Swap RGB/BGR');

    // Verify channel swap: bgr_test right half is stored as BGR-red ([0,0,255]) which
    // the extension displays as blue before swap (b>r), and red after swap (r>b).
    const imgAfterBgr = await debugHelper.captureCanvasImage();
    if (imgBeforeBgr && imgAfterBgr) {
      const rightBefore = sampleRegion(imgBeforeBgr, 0.55, 0.2, 0.35, 0.6);
      const rightAfter = sampleRegion(imgAfterBgr, 0.55, 0.2, 0.35, 0.6);
      DebugTestHelper.logger.info(`BGR swap right half — before: ${JSON.stringify(rightBefore)}, after: ${JSON.stringify(rightAfter)}`);
      assertChannelSwapped(rightBefore, rightAfter, 'b', 'r', 40, 'right region after Swap RGB/BGR');
    }

    DebugTestHelper.logger.success('All display options tests completed');
  }).timeout(300000);
});
